### PR TITLE
[MRG] Enable Py3 tests on travis

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
Python 2.7 was actually being used.

I suspect that travis is going to fail on this PR and that we will have to fix things for Python 3

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Wow, we got a segfault on travis :(.

Going to bed. We'll see what we can do about that tomorrow.

**Comment by GaelVaroquaux** [comment]
Tests pass on travis.

I have squashed all my commits. Please review this.

File: advanced/advanced_numpy/index.rst [hunk 1]
```diff
@@ -124,10 +124,20 @@ Reminder: two :class:`ndarrays <ndarray>` may share the same memory::
...
-    >>> x = '1234'
+    >>> x = '1234'      # In Python 3, use:   x = b'1234'
+
+.. For Python3
+   >>> try:
+   ...     x = b'1234'
+   ... except:
+   ...     pass
+
+x is a string (in Python 3 a buffer), we can represent it's data as an
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: its data

File: advanced/advanced_numpy/index.rst [hunk 2]
```diff
@@ -124,10 +124,20 @@ Reminder: two :class:`ndarrays <ndarray>` may share the same memory::
 
 Memory does not need to be owned by an :class:`ndarray`::
 
-    >>> x = '1234'
+    >>> x = '1234'      # In Python 3, use:   x = b'1234'
+
+.. For Python3
+   >>> try:
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: Why the try/except? ``x = b'1234'`` is valid Python 2 code, the 'b' is just ignored.

File: advanced/advanced_numpy/index.rst [hunk 3]
```diff
@@ -198,35 +207,45 @@ data_size          4-byte unsigned little-endian integer
...
+    
+    >>> try:
+    ...    str = bytes
+    ... except:
+    ...    skip
+
+The ``.wav`` file header as a Numpy *structured* data type::
+
+    >>> wav_header_dtype = np.dtype([
+    ...     ("chunk_id", (str, 4)),   # flexible-sized scalar type, item size 4
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: use ``bytes`` for Python 2.7 and Python 3 ?

Comment (GaelVaroquaux) [inline advanced/advanced_numpy/index.rst]: > use bytes for Python 2.7 and Python 3 ?

good suggestion!

File: advanced/advanced_numpy/index.rst [hunk 4]
```diff
@@ -249,15 +268,14 @@ dict_proxy({'block_align': (dtype('uint16'), 32), 'format': (dtype('S4'), 8), 'd
 
 >>> f = open('data/test.wav', 'r')
 >>> wav_header = np.fromfile(f, dtype=wav_header_dtype, count=1)
->>> f.close()
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: I don't understand why you removed this line (``f.close()``)

File: advanced/advanced_numpy/index.rst [hunk 5]
```diff
@@ -780,8 +800,8 @@ More tricks: diagonals
 
     >>> x = np.arange(5*5*5*5).reshape(5,5,5,5)
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: correct some pep8 while you're at it? (spaces)

File: advanced/advanced_numpy/index.rst (hunk 4 above)
Comment (GaelVaroquaux) [inline advanced/advanced_numpy/index.rst]: I could also skip it.

The problem is that in Python 3 it returns the number of bytes read (and not in Python 2)

File: advanced/advanced_numpy/index.rst [hunk 6]
```diff
@@ -780,8 +800,8 @@ More tricks: diagonals
...
     >>> s = 0
-    >>> for i in xrange(5):
-    ...    for j in xrange(5):
+    >>> for i in range(5):
+    ...    for j in range(5):
     ...       s += x[j,i,j,i]
```
Comment (emmanuelle) [inline advanced/advanced_numpy/index.rst]: same here, pep8

File: advanced/advanced_python/index.rst [hunk 7]
```diff
@@ -70,27 +70,27 @@ sequence allows us to have more than one way of iteration.
...
-2
->>> next(it)
-3
->>> next(it)
-Traceback (most recent call last):
-  File "<stdin>", line 1, in <module>
-StopIteration
+keystrokes. ::
+
+    >>> nums = [1,2,3]      # note that ... varies: these are different objects
```
Comment (emmanuelle) [inline advanced/advanced_python/index.rst]: pep8

File: advanced/advanced_python/index.rst [hunk 8]
```diff
@@ -148,9 +148,9 @@ braces. A ``dict`` is created when the generator expression contains
 If you are stuck at some previous Python version, the syntax is only a
 bit worse: ::
 
-    >>> set(i for i in 'abc')
+    >>> set(i for i in 'abc')   # doctest: +SKIP
```
Comment (emmanuelle) [inline advanced/advanced_python/index.rst]: ``{'a', 'b', 'c'}`` is returned by both Python 2.7 and Python 3 I think.

File: advanced/advanced_python/index.rst [hunk 9]
```diff
@@ -148,9 +148,9 @@ braces. A ``dict`` is created when the generator expression contains
...
     set(['a', 'c', 'b'])
-    >>> dict((i, ord(i)) for i in 'abc')
+    >>> dict((i, ord(i)) for i in 'abc')   # doctest: +SKIP
     {'a': 97, 'c': 99, 'b': 98}
 
 Generator expression are fairly simple, not much to say here. Only one
```
Comment (emmanuelle) [inline advanced/advanced_python/index.rst]: Generator expressions

File: intro/language/basic_types.rst [hunk 10]
```diff
@@ -64,9 +64,9 @@ Type conversion (casting)::
 
 .. warning:: Integer division
 
-    ::
+    In Python 2 (but not in Python 3)::
 
-        >>> 3 / 2
+        >>> 3 / 2   # doctest: +SKIP
```
Comment (emmanuelle) [inline intro/language/basic_types.rst]: write result for Python 3 as well ? (to be crystal clear...)

File: intro/language/control_flow.rst [hunk 11]
```diff
@@ -156,8 +155,10 @@ a file, ...)::
 
     >>> for i in 'powerful':
     ...     if i in vowels:
-    ...         print(i),
-    o e u
+    ...         print(i)
+    o
```
Comment (emmanuelle) [inline intro/language/control_flow.rst]: use ``print(i, end='')`` if you wish to keep the same layout as the example before. But I think it's fine as it is now.

File: intro/summary-exercises/optimize-fit.rst [hunk 12]
```diff
@@ -130,7 +130,7 @@ following arguments:
 
     >>> from scipy.optimize import leastsq
     >>> x, flag = leastsq(residuals, x0, args=(waveform_1, t))
-    >>> print x
+    >>> print(x)
     [  2.70363341  27.82020742  15.47924562   3.05636228]
```
Comment (emmanuelle) [inline intro/summary-exercises/optimize-fit.rst]: strange that this should not need an ellipsis...

**Comment by emmanuelle** [comment]
looks good to me, thanks heaps for the tedious but important work!

**Comment by GaelVaroquaux** [comment]
I've addressed all the comments, and the tests still pass. Merging! Thanks for the review.

**Pull request merged**